### PR TITLE
[KTO] fix various bugs

### PR DESCRIPTION
**Pull request by kawine**
PR includes following changes:

- make sure that the set of matched outputs y and the set of mismatched outputs y' (the latter for used the KL term) are the same within each batch (i.e., support of the empirical policy distribution should be consistent when calculating per-batch rewards and the KL term)
- fix issue in https://github.com/huggingface/trl/issues/1342 -- this was due to accelerate.gather not being able to properly handle anything that's not a 1x1 tensor
- add warnings to tell user to change the (un)desirable_weight hyperparams if the data is imbalanced

KTO vs DPO results now closer to what's in the paper. For Zephyr 7B:

KTO (~1ep):
MMLU: 58.2
GSM: 48.0
BBH: 52.0
TyDi: 40.7
HE: 35.1

DPO (1ep):
MMLU: 58.1
GSM: 18.5
BBH: 52.9
TyDi: 22.6
HE: 33.0

### CODE REVIEW

**Comment by kawine** [comment]
cc @kashif

**Comment by kashif** [comment]
thanks! for the timely fixes! reviewing now!

**Comment by kashif** [comment]
@kawine shall we inflate the test dummy datasets to avoid the exception there?

**Comment by kawine** [comment]
sure!

**Comment by kashif** [comment]
in the test can you try to return:

```
from datasets import concatenate_datasets


return concatenate_datasets([Dataset.from_dict(dummy_dataset_dict)] * 10)
```

File: trl/trainer/kto_trainer.py [hunk 1]
```diff
@@ -313,9 +313,16 @@ def make_inputs_require_grad(module, input, output):
         self.undesirable_weight = args.undesirable_weight
 
         # get KL datasets
-        train_KL_dataset = train_dataset.map(self.get_KL_dataset, batched=True, batch_size=1000)
+        total_batch_size = torch.cuda.device_count() * args.per_device_train_batch_size
```
Comment (lewtun) [inline trl/trainer/kto_trainer.py:316]: Would this be a more precise definition of total batch size?
```suggestion
        total_batch_size = torch.cuda.device_count() * args.per_device_train_batch_size * args.gradient_accumulation_steps
```

**Review by lewtun — APPROVED** [review]
Super nice fixes @kawine - thank you so much! I just left one nit and LGTM with Kashif's suggestions on the tests

File: trl/trainer/kto_trainer.py [hunk 2]
```diff
@@ -313,9 +313,16 @@ def make_inputs_require_grad(module, input, output):
         self.undesirable_weight = args.undesirable_weight
 
         # get KL datasets
-        train_KL_dataset = train_dataset.map(self.get_KL_dataset, batched=True, batch_size=1000)
+        total_batch_size = torch.cuda.device_count() * args.per_device_train_batch_size * args.gradient_accumulation_steps
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:316]:
```suggestion
        total_batch_size = (
             torch.cuda.device_count() * args.per_device_train_batch_size * args.gradient_accumulation_steps
         )
```

File: trl/trainer/kto_trainer.py [hunk 3]
```diff
@@ -313,9 +313,18 @@ def make_inputs_require_grad(module, input, output):
         self.undesirable_weight = args.undesirable_weight
 
         # get KL datasets
-        train_KL_dataset = train_dataset.map(self.get_KL_dataset, batched=True, batch_size=1000)
+        total_batch_size = (
+             torch.cuda.device_count() * args.per_device_train_batch_size * args.gradient_accumulation_steps
+         )
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:318]:
```suggestion
        total_batch_size = (
            max(torch.cuda.device_count(), 1) * args.per_device_train_batch_size * args.gradient_accumulation_steps
        )
```

**Comment by kawine** [comment]
thanks for wrapping this up @kashif ! sorry for being afk over the last couple days -- had some work to wrap up. everything looks good!

**Pull request merged**